pkg/reconciler/workspaceindex/server.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspaceindex
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sort"
+	"strconv"
+
+	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+)
+
+// NewServer creates a new server that can respond to requests for versioned data in workspaces.
+func NewServer(port int, waiter cacheSyncWaiter, index Index, stable func() bool) Server {
+	return &server{
+		port:   port,
+		waiter: waiter,
+		index:  index,
+		stable: stable,
+	}
+}
+
+type Server interface {
+	ListenAndServe(ctx context.Context)
+}
+
+type server struct {
+	port   int
+	waiter cacheSyncWaiter
+	index  Index
+	stable func() bool
+}
+
+type cacheSyncWaiter interface {
+	WaitForCacheSync(stopCh <-chan struct{}) map[reflect.Type]bool
+}
+
+func (s *server) ListenAndServe(ctx context.Context) {
+	mux := http.NewServeMux()
+	mux.Handle("/shard", http.HandlerFunc(s.handleShard))
+	mux.Handle("/data", http.HandlerFunc(s.handleData))
```

Comment:
Do we want to include an API version in the path, thinking about the future?

Reply:
I don't think necessarily that we need this API to even exist in the future, but it's a convenient way to test this in a black-box e2e.

Reply:
`/debug/data` is a better endpoint name, and used in many kube components.

Reply:
same for `/debug/shard`